helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java

```diff
@@ -202,6 +214,29 @@ protected void handleEvent(NotificationContext event) {
     }
   }
 
+  class RefreshTask implements Runnable {
+    @Override
+    public void run() {
+      try {
+        long currentTime = System.currentTimeMillis();
+        if (_lastEventTime + _periodicRefreshInterval <= currentTime) {
+          NotificationContext changeContext = new NotificationContext(_manager);
+          changeContext.setType(NotificationContext.Type.PERIODIC_REFRESH);
+          changeContext.setChangeType(_changeType);
+          enqueueTask(changeContext);
+        } else {
+          long remainingTime = _lastEventTime + _periodicRefreshInterval - currentTime;
+          _scheduledRefreshFuture.cancel(false);
+          _scheduledRefreshFuture = _periodicRefreshExecutor
+              .scheduleWithFixedDelay(this, remainingTime, _periodicRefreshInterval,
+                  TimeUnit.MILLISECONDS);
```

Comment:
Instead of doing this, can you just "sleep" for `remainingTime`, then do the check again? I think that automatically adjusts the schedule.

Reply:
Good idea! I changed it to sleep. One concern I had is that the thread could be interrupted while sleeping, but since we only have one thread here, I guess it's fine.

Reply:
There would be a good example of Delayed Scheduler. Try to understand that mode.

Reply:
This logic seems to be incorrect... But I guess you will change it anyway.

Reply:
Updated this part. Please take a look.